Add a cooldown to /report, resolves Electroid#242

applenick opened the pull request:
Resolves #242 

The changes were inspired by [here](https://github.com/OvercastNetwork/ProjectAres/blob/b883ef579938c49076379fa306592da847fc672f/Commons/bukkit/src/main/java/tc/oc/commons/bukkit/report/ReportCommands.java#L105) 

I did not want to deviate too far from what has been done in the past. The cooldown is implemented with a hard-coded 15 second cooldown, which I feel will be appropriate for most situations.

 If it is desired I can add a config option or adjust the hard-coded limit, just let me know.

~~Edit: Forgot to mention, I did not add a bypass for players with a specific permission. As I can’t really imagine any situation where one would want to bypass this check. Anyway, if that is desired I can happily add it too.~~ (Outdated)

Screenshot of formatting (what would happen if user spams the command several times):
![Screen Shot 2020-01-17 at 10 40 32 AM](https://user-images.githubusercontent.com/3377659/72637777-e3ac3900-3916-11ea-8a94-6926b390d8b0.png)
Signed-off-by: applenick <applenick@users.noreply.github.com>

Meeples10 commented:
There should probably be a config option for this that defaults to 15 seconds.

Pablete1234 commented on src/main/java/tc/oc/pgm/commands/ModerationCommands.java:
```diff
@@ -18,16 +24,40 @@
 
+  private static final int REPORT_COOLDOWN_SECONDS = 15;
+
+  private final Map<CommandSender, Instant> lastReportSent = new WeakHashMap<>();
```
Should use a guava expiring cache, and maybe just use the UUID as key (if the command sender is the console just always allow)

Electroid replied:
Also exempt players with the `Permissions.STAFF`

applenick replied:
Thanks @Pablete1234 and @Electroid! 
I switched the  map over to use a guava loading cache. I haven’t used them before, so hopefully the way I implemented it is the proper way. Though after testing everything seemed to work, so it should be right 😄 

Also I added the staff bypass.

Pablete1234 commented:
I think the 15 second constant is fine, there's no need to over-complicate such a simple feature

applenick commented:
Thanks @Pablete1234 and @Electroid!
I switched the map over to use a guava loading cache. I haven’t used them before, so hopefully the way I implemented it is the proper way. Though after testing everything seemed to work, so it should be right 😄

I also agree with @Pablete1234 that this is a very simple feature and really there is no reason to add an additional config option.

Pablete1234 commented on src/main/java/tc/oc/pgm/commands/ModerationCommands.java:
```diff
@@ -18,6 +26,14 @@
+
+  private static final Cache<UUID, Instant> lastReportSent =
+      CacheBuilder.newBuilder()
+          .weakKeys()
```
Using uuis, no need for weak keys

applenick replied:
Ah thanks, I was unsure! I’ll adjust that now

Pugzy commented:
Would it be better to make the translated message more generic?
> Please wait {0} before running that command again

Allowing this to be extracted to more commands in the future

applenick commented:
> Would it be better to make the translated message more generic?
> 
> > Please wait {0} before running that command again
> 
> Allowing this to be extracted to more commands in the future

Great suggestion. I’ll adjust that as well.

TheMolkaPL commented on src/main/java/tc/oc/pgm/commands/ModerationCommands.java:
```diff
@@ -27,7 +40,32 @@ public static void report(
+                      ChatColor.AQUA
+                          + (secondsRemaining
+                              + " second"
+                              + (secondsRemaining != 1 ? "s" : "")
```
This should be translatable.

applenick replied:
Thanks Molka, I should have done this from the beginning!

TheMolkaPL commented:
I meant something like this: :P

https://github.com/Electroid/PGM/blob/31ebbcc74002981fcf05dc0af45e861a77d07aca/src/main/i18n/templates/strings.properties#L672-L674

TheMolkaPL commented on src/main/java/tc/oc/pgm/commands/ModerationCommands.java:
```diff
@@ -18,6 +25,11 @@
+  private static final int REPORT_COOLDOWN_SECONDS = 15;
+
+  private static final Cache<UUID, Instant> lastReportSent =
+      CacheBuilder.newBuilder().expireAfterWrite(REPORT_COOLDOWN_SECONDS, TimeUnit.SECONDS).build();
```
I think this should be called `LAST_REPORT_SENT`.

applenick replied:
> I meant something like this: :P
> 
> https://github.com/Electroid/PGM/blob/31ebbcc74002981fcf05dc0af45e861a77d07aca/src/main/i18n/templates/strings.properties#L672-L674

Whoops! 😆 
I’ll make the adjustments now, thanks again haha.

TheMolkaPL commented on src/main/java/tc/oc/pgm/commands/ModerationCommands.java:
```diff
@@ -27,7 +39,36 @@ public static void report(
+        Duration timeSinceReport = Duration.between(lastReport, Instant.now());
+        long secondsRemaining = REPORT_COOLDOWN_SECONDS - timeSinceReport.getSeconds();
+        if (secondsRemaining > 0) {
+          Component secondsComponent = new PersonalizedText("" + secondsRemaining);
```
I would use `Integer.toString(secondsRemaining)`, otherwise looks good to me.

applenick replied:
Oh snap, I did not see your suggestion! I’m adjusting this right now. 👍
